File: src/nodes/builtin_func_call.zig
const std = @import("std");
const nd = @import("node.zig");
const bytecode = @import("../common/bytecodes.zig");

pub const NodeBuiltinFuncCall = struct {
    func_bytecode: bytecode.ApicaBuiltinFuncCallBytecode,
    parameters: std.ArrayList(*nd.Node),

    pub fn init(func_bytecode: bytecode.ApicaBuiltinFuncCallBytecode, params: std.ArrayList(*nd.Node)) NodeBuiltinFuncCall {
        return NodeBuiltinFuncCall{ .func_bytecode = func_bytecode, .parameters = params };
    }

    pub fn get_func_bytecode(self: *const NodeBuiltinFuncCall) bytecode.ApicaBuiltinFuncCallBytecode {
        return self.func_bytecode;
    }

    pub fn get_parameters(self: *const NodeBuiltinFuncCall) std.ArrayList(*nd.Node) {
        return self.parameters;
    }
};

File: src/nodes/builtin_func_call.rs
use apica_common::bytecodes::ApicaBuiltinFunctionBytecode;
use crate::nodes::node::Node;

pub struct NodeBuiltinFuncCall {
    func_bytecode: ApicaBuiltinFunctionBytecode,
    parameters: Vec<Node>,
}

impl NodeBuiltinFuncCall {
    pub fn init(func_bytecode: ApicaBuiltinFunctionBytecode, parameters: Vec<Node>) -> NodeBuiltinFuncCall {
        NodeBuiltinFuncCall { func_bytecode, parameters }
    }
    
    pub fn get_function_bytecode(&self) -> &ApicaBuiltinFunctionBytecode {
        &self.func_bytecode
    }
    
    pub fn get_parameters(&self) -> &Vec<Node> {
        &self.parameters
    }
}